Sdma cleanup

paulstelian97 opened the pull request.

Commit:
Debug aids

Commit:
drivers: imx: interrupt: Adjust IRQ_STEER registers for i.MX8

Signed-off-by: Paul Olaru <paul.olaru@nxp.com>

Commit:
[DO NOT MERGE] platform: imx8m: Add more heap memory

The extra memory blocks are required for SDMA proper functioning.
TODO figure out which of the increases are actually required.

Signed-off-by: Paul Olaru <paul.olaru@nxp.com>

Commit:
[DO NOT MERGE] platform: imx8m: Add SDMA definition

This commit adds the definition of the SDMA controller we use for SOF,
namely SDMA2 (the other instances of SDMA are used for non-audio-related
DMA in the system). This DMAC can serve the SAI.

[DO NOT MERGE] because I need help in how to account for channel 0 being
used for management functionality, with only channels 1-31 being usable
by SOF directly.

Signed-off-by: Paul Olaru <paul.olaru@nxp.com>

Commit:
platform: imx8m: sdma: Add SDMA definitions for i.MX8M platform

This includes interrupt number and a tweak used when probing the driver
(which must match hardware configuration).

Signed-off-by: Paul Olaru <paul.olaru@nxp.com>

Commit:
platform: imx8m: Fix DMA domain initialization

SDMA is in dma_array[1], and uses a single interrupt for all channels.

Signed-off-by: Paul Olaru <paul.olaru@nxp.com>

dbaluta commented on src/platform/imx8m/lib/dai.c:
```diff
@@ -18,6 +18,17 @@ static SHARED_DATA struct dai sai[] = {
 	.index = 3,
 	.plat_data = {
 		.base = SAI_3_BASE,
```
This can go as a separate PR. Please add some more info in the commit message. The current commit message is not very informative.

paulstelian97 replied:
Will do.

dbaluta commented on src/platform/imx8m/lib/dai.c:
```diff
@@ -18,6 +18,17 @@ static SHARED_DATA struct dai sai[] = {
+			.offset = SAI_3_BASE + REG_SAI_RDR0,
+			.depth = 16,
+			.handshake = 4,
+		},
```
Maybe we could add some meaningful macros for event 4 and 5.

dbaluta commented on src/platform/imx8m/lib/dai.c:
```diff
@@ -18,6 +18,17 @@ static SHARED_DATA struct dai sai[] = {
+		.fifo[SOF_IPC_STREAM_PLAYBACK] = {
+			.offset = SAI_3_BASE + REG_SAI_TDR0,
+			.depth = 16,
+			/* Handshake is SDMA hardware event */
```
Looking in the upstream SAI Linux driver (sound/soc/fsl/fsl_sai.c) depth for i.MX8MQ is 128. I think that SAI is the same on 8MP vs 8MP.

paulstelian97 replied:
Depth is in number of words and is interpreted by SDMA driver. Will need to check.

dbaluta commented on src/platform/imx8m/include/platform/lib/dma.h:
```diff
@@ -21,6 +21,17 @@
+
+#define SDMA2_IRQ	7 /* TODO What? */
+#define SDMA2_IRQ_NAME	"irqstr2" /* TODO find the correct one */
+
```
Whe should remove the TODOs here. If we are sure the numbers are correct.

paulstelian97 replied:
I am not 100% certain. I may also end up using irqstr_get_sof_int like on the other platforms (perhaps my calculations were in fact incorrect).

dbaluta commented on src/platform/imx8m/lib/dma.c:
```diff
@@ -12,6 +12,7 @@
 #include <sof/spinlock.h>
 
 extern struct dma_ops dummy_dma_ops;
```
Here we should update the commit message once all questions are answered.

dbaluta commented on src/platform/imx8m/include/platform/lib/dma.h:
```diff
@@ -15,7 +15,7 @@
 
-#define DMA_ID_EDMA0	0
+#define DMA_ID_SDMA2	0
 #define DMA_ID_HOST	1
```
good catch. I think we can make the following PRs:

1) PR with cleanups in preparation for SDMA driver.
2) PR adding SDMA driver

We could get 1) accepted faster so that i.MX8MP boots fine with qemu.

dbaluta commented on src/drivers/imx/CMakeLists.txt:
```diff
@@ -6,6 +6,7 @@ add_local_sources(sof
 	timer.c
+	sdma.c
 )
 
```
TODO: Add selectable config for SDMA. e.g: CONFIG_IMX_SDMA. This can be done later.

paulstelian97 replied:
I think I can do this in the same PR that will have the implementation.

dbaluta commented on src/drivers/imx/CMakeLists.txt:
```diff
@@ -6,6 +6,7 @@ add_local_sources(sof
 	interrupt.c
```
We could cleanup commit message, be deleting DO NOT MERGE and this commit requires the cleanup. Also, would be nice to add a link to a RM containing SDMA bits. Maybe link to 8MQ RM.

paulstelian97 replied:
That's a reminder for myself that I need to clean up some bits from this commit. We have seen how unreliable direct links for the RM are so I will look into how to make a reasonably-persistent link for future reference.

dbaluta commented on src/drivers/imx/sdma.c:
```diff
@@ -0,0 +1,848 @@
+
+struct sdma_bd {
+	/* SDMA BD configuration */
+	uint32_t config;
```
maybe first time use full name for BD, not sure what it stands for.

dbaluta commented on src/drivers/imx/sdma.c:
```diff
@@ -0,0 +1,848 @@
+	trace_sdma_error("sdma_set_overrides(%d, %d)", event_override, host_override);
+	dma_reg_update_bits(channel->dma, SDMA_EVTOVR, BIT(channel->index), event_override ? BIT(channel->index) : 0);
+	dma_reg_update_bits(channel->dma, SDMA_HOSTOVR, BIT(channel->index), host_override ? BIT(channel->index) : 0);
+}
```
I think this line goes over 80 chars. In fact I think there are several places where this happens. Maybe run checkpatch.ppl with --strict?

paulstelian97 replied:
Will fix the lines-too-long issues as I look again into this file. I still have a few dummy functions that need to go.

dbaluta replied:
thanks.

dbaluta commented on src/drivers/imx/sdma.c:
```diff
@@ -0,0 +1,848 @@
+
+	/* Writeback descriptors and CCB */
+
+	dcache_writeback_region(c0data->descriptors, sizeof(c0data->descriptors[0]));
```
No need for new line here.

dbaluta commented on src/drivers/imx/sdma.c:
```diff
@@ -0,0 +1,848 @@
+	sdma_set_overrides(c0, true, false);
+	ret = dma_start(&dma->chan[0]);
+
+	if (ret < 0)
```
Same here, no need for new line.

dbaluta commented on src/drivers/imx/sdma.c:
```diff
@@ -0,0 +1,848 @@
+
+	trace_sdma("SDMA: probe");
+	trace_sdma_error("sdma_probe");
+
```
Just keep trace_sdma here.

paulstelian97 replied:
The trace_sdma_error calls will all go (some may be converted to trace or trace_verbose if they are important enough)

dbaluta replied:
thanks.

dbaluta commented on src/drivers/imx/sdma.c:
```diff
@@ -0,0 +1,848 @@
+	}
+	trace_sdma_error("sdma_set_event(%d, %d)", channel->index, eventnum);
+	struct sdma_chan *pdata = dma_chan_get_data(channel);
+	dma_reg_update_bits(channel->dma, SDMA_CHNENBL(eventnum), BIT(channel->index), BIT(channel->index));
```
I would suggest on declaring all the variables at the beginning of function.

paulstelian97 replied:
Will do.

dbaluta commented on src/drivers/imx/sdma.c:
```diff
@@ -0,0 +1,848 @@
+	for (int i = 0; i < dma->plat_data.channels; i++) {
+		struct dma_chan_data *channel = &dma->chan[i];
+		struct sdma_chan *cdata = &pdata->chan_pdata[i];
+
```
Declare all the variable at the beginning of function. Hope this wont trigger an compiler error for unused variable.

dbaluta commented on src/drivers/imx/sdma.c:
```diff
@@ -0,0 +1,848 @@
+		return -EINVAL;
+	}
+	bool src_may_change, dst_may_change;
+	switch (config->direction) {
```
move declaration at the beginning of function.

dbaluta commented on src/include/sof/drivers/sdma.h:
```diff
@@ -0,0 +1,152 @@
+#define SDMA_SCRIPT_SHP2MCU_OFF		979
+#define SDMA_SCRIPT_MCU2SHP_OFF		1048
+
+#endif /* __SOF_DRIVERS_SDMA_H__ */
```
Other than that it looks good.

Commit:
drivers: imx: sdma: Add SDMA support to SOF

Some i.MX platforms use SDMA (Smart DMA) controller for DMA related
operations. This commit adds the initial functionality for SDMA.

Signed-off-by: Paul Olaru <paul.olaru@nxp.com>
